### PR TITLE
Superior type inference

### PR DESCRIPTION
**Pull request by paulp**
There are some long commit messages - I refer you to those. Review by @retronym, @adriaanm.

### CODE REVIEW

File: src/compiler/scala/tools/nsc/typechecker/Infer.scala [hunk 1]
```diff
@@ -1120,127 +1162,105 @@ trait Infer extends Checkable {
...
+      def patternFirst = pattpKnown && !ptKnown
+
+      // The expected type, as seen from the pattern's type symbol, after taking into account
+      // that a value exists which must conform to both types.
+      def ptPropagated = propagateKnownTypes(pt, pattp.widen.typeSymbol)
+
+      // If we can absolutely rule out a match we can fail early.
+      // This is possibly the case if the scrutinee has no unresolved type arguments
+      // and is a "final type", meaning final + invariant in all type parameters.
+      def isIncompatible = pt.isFinalType && ptKnown && !(ptWild matchesPattern pattp.widen)
```
Comment (retronym) [inline src/compiler/scala/tools/nsc/typechecker/Infer.scala:1208]: I'm marking out what seem like interesting parts of the diff to highlight them from the refactorings. `&& ptKnown` is added here, and `ptWild matchesPattern` used to be `pt matchesPattern`.

File: src/compiler/scala/tools/nsc/typechecker/Infer.scala [hunk 2]
```diff
@@ -1120,127 +1162,105 @@ trait Infer extends Checkable {
...
+      // Generally the inferred type will be `pt with pattp`, which means the pattern
+      // type is stronger than the expected type. This is reversed in cases where the pattern
+      // type has no free type parameters, but the expected type does. See pos/t2486 and
+      // pos/channels for some examples of when this is necessary.
+      if (isIncompatible)
+        andError(IncompatibleScrutineeTypeError(tree0, pattp, pt))
+      else if (conformsDirectly)
+        pattp
+      else if (conformsAdapted)
+        intersection
```
Comment (retronym) [inline src/compiler/scala/tools/nsc/typechecker/Infer.scala:1240]: The crux: we get here more often thanks to `propagateKnownTypes`. Before:

``` scala
          if (isPopulated(tp, pt1) && isInstantiatable(tvars ++ ptvars) || pattpMatchesPt)
             ptvars foreach instantiateTypeVar
          else {
            PatternTypeIncompatibleWithPtError1(tree0, pattp, pt)
            return ErrorType
```

File: src/compiler/scala/tools/nsc/typechecker/Infer.scala [hunk 3]
```diff
@@ -173,10 +199,39 @@ trait Infer extends Checkable {
   private lazy val stdErrorClass = rootMirror.RootClass.newErrorClass(tpnme.ERROR)
   private lazy val stdErrorValue = stdErrorClass.newErrorValue(nme.ERROR)
 
+  /** Create fresh TypeVars for `tparams` and instantiate `tp` with those TypeVars as
+   *  its type arguments. Finally call `fn` with the type and type vars.
+   */
+  def withFreshVars[T](tp: Type, tparams: List[Symbol])(fn: (Type, List[TypeVar]) => T): T = {
```
Comment (retronym) [inline src/compiler/scala/tools/nsc/typechecker/Infer.scala:205]: Wonderful!

**Comment by huitseeker** [comment]
Those IDE validation failures ([bc9630e](https://scala-webapps.epfl.ch/jenkins/view/pr-validators/job/pr-scala-integrate-ide/1672/console) [7c8b636](https://scala-webapps.epfl.ch/jenkins/view/pr-validators/job/pr-scala-integrate-ide/1668/console) ) look genuine (but not surprising).

File: src/compiler/scala/tools/nsc/typechecker/Infer.scala [hunk 4]
```diff
@@ -1120,127 +1162,105 @@ trait Infer extends Checkable {
...
+    private def instantiateIfPossible(tvars: List[TypeVar]) = (
+         isSolvable(tvars map (_.cloneInternal))
+      && instantiateTypeVars(tvars)
+    )
+    private def andError(body: Unit): Type = ErrorType
+
     /** Type intersection of simple type tp1 with general type tp2.
      *  The result eliminates some redundancies.
      */
     def intersect(tp1: Type, tp2: Type): Type = {
```
Comment (retronym) [inline src/compiler/scala/tools/nsc/typechecker/Infer.scala:1179]: I wonder if we should rename this to convey the assumptions we making by using `propagateKnownTypes`.

It's only used in `inferTypedPattern` right now; we could also make it private.

File: src/compiler/scala/tools/nsc/typechecker/Checkable.scala [hunk 5]
```diff
@@ -49,41 +49,42 @@ trait Checkable {
   import definitions._
   import CheckabilityChecker.{ isNeverSubType, isNeverSubClass }
 
+  case class TypeInference(tycon: Type, tparams: List[Symbol], tvars: List[TypeVar]) {
+    private def instTypeArgs = tparams zip tvars map {
+      case (_, tvar) if tvar.instValid => tvar.constr.inst
+      case (tparam, _)                 => tparam.tpeHK
+    }
+    def solve = appliedType(tycon, instTypeArgs)
```
Comment (retronym) [inline src/compiler/scala/tools/nsc/typechecker/Checkable.scala:57]: @adriaanm Perhaps you'll have some insight to offer here. Last time Paul and I were discussing `propagateKnownTypes` on the mailing lists:

File: src/compiler/scala/tools/nsc/typechecker/Checkable.scala [hunk 6]
```diff
@@ -49,41 +49,42 @@ trait Checkable {
   import definitions._
   import CheckabilityChecker.{ isNeverSubType, isNeverSubClass }
 
+  case class TypeInference(tycon: Type, tparams: List[Symbol], tvars: List[TypeVar]) {
```
Comment (retronym) [inline src/compiler/scala/tools/nsc/typechecker/Checkable.scala:52]: I don't see the point of making this a class, it seems like a method would do.

File: src/reflect/scala/reflect/internal/tpe/TypeMaps.scala [hunk 7]
```diff
@@ -53,6 +54,24 @@ private[internal] trait TypeMaps {
     }
   }
 
+  // TODO: this is extremely rough...
+  // replace type args by wildcards, since they can't be checked (don't use existentials: overkill)
+  // TODO: when type tags are available, we will check -- when this is implemented, can we take that into account here?
+  // similar to typer.infer.approximateAbstracts
+  object typeArgsToWildcards extends TypeMap {
+    // SI-6771 dealias would be enough today, but future proofing with the dealiasWiden.
+    // See neg/t6771b.scala for elaboration
```
Comment (retronym) [inline src/reflect/scala/reflect/internal/tpe/TypeMaps.scala:63]: Happily, this comment is now out of date!

File: src/reflect/scala/reflect/internal/tpe/TypeMaps.scala [hunk 8]
```diff
@@ -53,6 +54,24 @@ private[internal] trait TypeMaps {
...
+    // Note that this depends on NOT traversing the prefix, else nested classes
+    // which depend on outer pointer semantics (SI-4440) will break.
+    def apply(tp: Type): Type = tp.dealias match {
```
Comment (retronym) [inline src/reflect/scala/reflect/internal/tpe/TypeMaps.scala:66]: This was `.dealiasWiden`. I need to poke things a bit to see if that change makes sense.

In my poking, I've found a regression. It's not related to this line though (I tried reverting and it still failes)

```
object Test {

  val a: Option[Int] = None

  (a: a.type) match {
     case Some(x) => println(x)
     case None    =>
  }
}
```

```
% scalac-hash v2.11.0-M4 test/files/pos/t6771-singleton.scala
[info] v2.11.0-M4 => /Users/jason/usr/scala-v2.11.0-M4-0-g8b41240

% qbin/scalac test/files/pos/t6771-singleton.scala
test/files/pos/t6771-singleton.scala:6: error: constructor cannot be instantiated to expected type;
 found   : Some[A]
 required: Test.a.type
     case Some(x) => println(x)
          ^
test/files/pos/t6771.scala:6: error: not found: value x
     case Some(x) => println(x)
                             ^
two errors found

```

Comment (retronym) [inline src/reflect/scala/reflect/internal/tpe/TypeMaps.scala:66]: Similarly:

```
object Test {

  val a: Option[Int] = None

  (a: a.type) match {
     case s: Some[Int] => println(s)
     case None    =>
  }
}
```

```
% qbin/scalac test/files/pos/t6771.scala
test/files/pos/t6771.scala:6: error: type mismatch;
 found   : Some[Int]
 required: Test.a.type
     case s: Some[Int] => println(s)
           ^
one error found
```

Comment (retronym) [inline src/reflect/scala/reflect/internal/tpe/TypeMaps.scala:66]: Maybe it's best to disentangle the changes a bit: a) refactoring of type inference boilerplate, b) sharper intersection types,  c) propagation of singleton types.

**Comment by paulp** [comment]
@huitseeker how is anyone supposed to identify the "genuine" failure in all the megabytes of noise?

**Comment by adriaanm** [comment]
I'll rebase this when I get back from vacation in two weeks.

**Comment by retronym** [comment]
I'm working on a rebased version of the PR, that includes the extra tests from my review comments.

https://github.com/retronym/scala/tree/topic/2894-rebase

File: src/compiler/scala/tools/nsc/typechecker/Infer.scala [hunk 9]
```diff
@@ -1120,127 +1162,105 @@ trait Infer extends Checkable {
...
-        val reduced2 = tp2 match {
-          case rtp @ RefinedType(parents2, decls2) =>
-            copyRefinedType(rtp, parents2 filterNot (tp1 <:< _), decls2)
-          case _ =>
-            tp2
-        }
-        intersectionType(List(tp1, reduced2))
-      }
+      else if (sym1 isStrictSubClass sym2) logResult(s"(1) intersect($tp1, $tp2)")(intersect(tp1, propagateKnownTypes(tp2, sym1)))
+      else if (sym2 isStrictSubClass sym1) logResult(s"(2) intersect($tp1, $tp2)")(intersect(propagateKnownTypes(tp1, sym2), tp2))
```
Comment (retronym) [inline src/compiler/scala/tools/nsc/typechecker/Infer.scala:1190]: Singleton types are dropped on the floor here:

``` scala
scala> val a: Option[Int] = null
a: Option[Int] = null

scala> typer.infer.intersect(typeOf[a.type], typeOf[Option[Int]])
res1: $r.intp.global.analyzer.global.Type = a.type

scala> typer.infer.intersect(typeOf[a.type], typeOf[Some[Int]])
res2: $r.intp.global.analyzer.global.Type = Some[Int]
```

Comment (retronym) [inline src/compiler/scala/tools/nsc/typechecker/Infer.scala:1190]: Note to self: what about existential pts?

**Pull request closed without merging**